### ISSUE TITLE
[ Feature Request ] - FileSystem<T> 

### ISSUE THREAD

**Issue by ShanonJackson** [microsoft/TypeScript#42046]
## Suggestion
```
FileSystem<T> type

Works with alot of simplicity
- If it hits a folder adds it into the 'object' type and its children as traversable keys
- If it hits a file thats a module it adds whatever its ```tstypeof import("..")``` type would be.
- Recursive
```

## Use Cases and Examples

**1:**  Every use case where previously people had alot of types or classes floating around and then had a single place where they imported all those types/classes into a union just to support typescript can now just use a folder to store those things in conjunction with `FileSystem`
```ts
// people using ORM's can avoid having a 'single' source where all their classes are imported in favor of just using a folder.
await database.user(123);

// implementation 
class SomeORMClass {
    getUsers() {
        return [];
    }
}

// Replaced via FileSystem helper type.
type FileSystem = {
    users: {default: SomeORMClass} // representation of import("..") that has a default class export
}

type ORM = {
    [K in keyof FileSystem]: FileSystem[K]["default"]
}

declare const database: ORM ;
database.users.getUsers();
```

**2:** Really strong support for image paths, and removing files in refactors that are referenced by string path.
```ts
 // (note: requires augmentation of JSX types);
<img src={"/static/dolphin.png"}/> // error dolphin.png isn't loaded in the file system.
```

**3:**  Support for string based routing in applications and protection against routing refactors at a type-level.
```ts
 // (note: requires augmentation of JSX types);
<a href={"/home/name"}/>

// currently people do something like this just to have 'route refactorability' which bloats bundle size.
const Routes = {
     Home: "/"
} as const
```

**4:**  Statically typed readFileSync for people using it to access their file system
```ts
const readFileSystem = (path: keyof FileSystem<".">): string=> fs.readFileSync(path);
```

**5:**  Statically typed API handling
```ts
(pseudo code)
// before
const response = await api<typeof import("/api/user")>("/api/user")
// after
const response = await api.user(1);
// implementation
const api = new Proxy<FileSystem<"../../../api">>(...);
```

+ Many more use cases (File system based redux, file system based statically typed ExpressJS etc)

**Precedence:** When people look at the implementation in the standard library they will notice FileSystem doesn't have a implementation in the same way you can "peek definition" for types like `Exclude` and `Omit`, however there already is precedence for type-level helpers that work as a 'compiler flag' (I.E `ThisType<T>`);

Previous issue:
https://github.com/microsoft/TypeScript/issues/41550

Further important notes:

- People are going to want to type things with the 'flattened' keys of a FileSytem<T> in various formats which will easily be possible to use it in conjunction with existing mapped types and new string literal types.
- People are going to want to use FileSystem to represent things that are FileSystem'Like which again is easily possible with mapped/string literal
- People are going to want to use FileSystem to represent transformations that could happen at build time which is again possible with mapped/string literal types.

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by pauldraper** [comment]
FileSystem is the wrong name IMO.

Should be "Module" or "Import".

**Comment by ShanonJackson** [comment]
I'm terrible at naming things, and I've noticed the typescript core team seems to be very good. So I'm definitely open to suggestions however I would like this name to both be something that represents "many" or plural in some way, and has some meaning relating to a folder/file structure.

Module - Not plural, currently already has a different meaning in Javascript, has no meaning relating to folder/file structures.
Import - Not plural, currently already has a different meaning in Javascript, is already a used word syntactically in typescript and has some contextual meaning relating to folder/file structures but as individual files/folders not many.


I'm happy for a name that involves concise reasoning what I do like about Module/Imports is this type is definitely related to those things, however not represented by those things alone.

**Comment by pauldraper** [comment]
The naming comes from: "Same semantics as import(".....") type."

`import` itself knows nothing about files, directories, URLs, etc; it just hooks into TypeScript module resolution (which translates module specifier to module).

```
declare module "zoomdiddlywhop//$1!01#x=1>" {
    export const example: number;
}

const a = import("zoomdiddlywhop//$1!01#x=1>");
```

`Module<ModuleId>`, `Import<ModuleId>`, `typeof import` conveys that this is the type imported by `import`.

**Comment by phenomnomnominal** [comment]
I'm a big fan of `Import<ModuleId>`. With the `FileSystem` name it wasn't even clear to me that this suggestion was the same thing that I wanted.

**Comment by thetutlage** [comment]
What are the chances of this ever getting implemented? Asking coz there are similar opened issues for the samehttps://github.com/microsoft/TypeScript/issues/31090

I have alternative, more verbose ways to achieve it, but still waiting to see if there is any possibility of this getting implemented

**Comment by RyanCavanaugh** [comment]
> What are the chances of this ever getting implemented?

I try not to say "never" on things that aren't clearly out-of-scope, but the complexity issues here are *massive* since we're dealing with the file system, which is much more than just a simple tree. Issues I would expect us to encounter immediately:
 * Which files are part of this, and which aren't? Is it only the files that are otherwise in the program, or any file we can find? From an architectural perspective, we can only implement the former. Either answer is going to surprise and annoy at least 30% of users and prompt a request for a commandline flag or other configuration path.
 * In rare cases in TypeScript, order matters, so file enumeration order from the OS would either be a potential build break, or we'd have to normalize on some canonical sort order (probably fine)
 * Are there going to be feature requests for reconfiguring the magic string `'index'`? Seems valid since there are not necessarily commons modules
 * Is `FileSystem<notfoundpath>` an error, or a no-op? Either answer is going to surprise and annoy at least 30% of users and prompt a request for a commandline flag or other configuration path.
 * How does this interact with case-insensitive file systems? If we're both on case-insensitive systems, then I can have `A.ts` and you can have `a.ts` and we're both OK, but only one of us can be right on the type side
 * It seems like this is only useful on the type side, but presumably you'd want to have a value-side view of this same thing for the same reasons of convenience, which makes it seem like it should just be behavior provided by hosting module system?

Basically the upside value here when compared to running an out-of-band tool to autogen a .d.ts file when the file layout changed would need to be very, very, *very* compelling for us to open the floodgates on the bug farm and configuration Rubik's cube this would introduce.

**Comment by pauldraper** [comment]
As I understand it, it is a duplicate of #31090 .

And with that understanding lot of @RyanCavanaugh 's questions go away (we're not talking about OS files, but rather modules).

**Comment by thetutlage** [comment]
As @pauldraper mentioned, the main request here is to have `Import` as a type that accepts a generic alongside the `import` keyword. I think, the FileSystem is a wrong name all together.

**Comment by ShanonJackson** [comment]
> * Which files are part of this, and which aren't? Is it only the files that are otherwise in the program, or any file we can find? From an architectural perspective, we can only implement the former. Either answer is going to surprise and annoy at least 30% of users and prompt a request for a commandline flag or other configuration path.

I would say we aim for identical behaviour as  ```typeof import("")``` where possible, so if you want png's to appear, you have to augment etc.


> * In rare cases in TypeScript, order matters, so file enumeration order from the OS would either be a potential build break, or we'd have to normalize on some canonical sort order (probably fine)

Canonical sort order will work but I'm open to suggestions


> * Are there going to be feature requests for reconfiguring the magic string `'index'`? Seems valid since there are not necessarily commonjs modules

Open to suggestions but I don't think this a deal breaker we can potentially use intersections on the root instead of adding a key with "index", not too fimiliar with how the different module systems handle this but I'd like your guidance on how we should handle this @RyanCavanaugh .

Is FileSystem<notfoundpath> an error, or a no-op? Either answer is going to surprise and annoy at least 30% of users and prompt a request for a commandline flag or other configuration path.

Same behaviour as  ```typeof import("")``` so yes it will be an error, but this both keeps it consistent and annoys the same subset of people.

> * How does this interact with case-insensitive file systems? If we're both on case-insensitive systems, then I can have `A.ts` and you can have `a.ts` and we're both OK, but only one of us can be right on the type side

Not familiar with this, but again I would say ```typeof import()``` already has to handle this case so we attempt to do it same way, if we can't open to suggestions but I don't think this is unsolvable. 


> * It seems like this is only useful on the type side, but presumably you'd want to have a value-side view of this same thing for the same reasons of convenience, which makes it seem like it should just be behavior provided by hosting module system?

Wont to keep the scope limited for now, I think there is immense value in this proposal for the use cases listed here + many more, all of the above examples can purely be solved with a type-level implementation. I think in the future a value-side view could be possible.
